Add syntax for kinds

ewang12 opened the pull request:
Add kinds.

[Here](https://s3.amazonaws.com/stephan-misc/paper/branch-syntax-kinds.pdf) is a link to the PDF generated from this PR.

stepchowfun commented on main.tex:
```diff
@@ -107,23 +116,28 @@
-          & $\eeffect{\xtapp{\xeffect}{\tvar}}{\evar}{\tx}{\eterm}$ & effect declaration \\
-          & $\eprovide{\xtapp{\xeffect}{\tx}}{\xeffects}{\evar}{\eterm}{\eterm}$ & effect definition \\
+          & $\eeffect{\xtapp{\xeffect}{\lstof{\tanno{\tvar}{\kkind}}}}{\evar}{\tx}{\eterm}$ & effect declaration \\
+          & $\eprovide{\xtapp{\tx}}{\xeffects}{\evar}{\eterm}{\eterm}$ & effect definition \\
```
`\xtapp` takes two arguments ~

stepchowfun replied:
(i think let's just delete it)

stepchowfun commented on main.tex:
```diff
@@ -107,23 +116,28 @@
           & $\tarrow{\tx}{\tx}$ & arrow type \\
-          & $\ttforall{\tvar}{\tx}$ & quantified type \\
-          & $\txforall{\xvar}{\tx}$ & quantified effect \\
           $\tx \Coloneqq$ & & type with effects \\
```
i think instead of `type with effects`, we should come up with a more accurate description of this thing. but i don't know what!

stepchowfun commented on main.tex:
```diff
@@ -107,23 +116,28 @@
-          & $\txforall{\xvar}{\tx}$ & quantified effect \\
           $\tx \Coloneqq$ & & type with effects \\
           & $\twithx{\ttype}{\xeffects}$ & effect row bound \\
+          & $\tvar$ & type variable \\
```
technically could also be used for effects, not just types, right?

stepchowfun commented on main.tex:
```diff
@@ -107,23 +116,28 @@
+          & $\tvar$ & type variable \\
+          & $\ttabs{\tanno{\tvar}{\kkind}}{\tx}$ & operator abstraction \\
+          & $\ttapp{\tx}{\tx}$ & operator application \\
+          & $\xeffect$ & effect variable \\
```
i'm wondering if this counts as a "variable" (cf. `\tvar`) ~ what to you think

ewang12 replied:
It seems like `\tvar` is introduced in an operator abstraction, whereas an effect variable is introduced by an effect declaration. But I don't know if that matters.

stepchowfun commented on main.tex:
```diff
@@ -107,23 +116,28 @@
+          & $\xeffect$ & effect variable \\
+          & $\xeffects$ & effect row \\
+          $\kkind \Coloneqq $ & & kind \\
+          & $\ktype$ & type kind \\
```
i think this almost sounds like an oxymoron, though it makes sense. but maybe we should use pierce's language ("kind of proper types")

ewang12 replied:
I thought maybe that wouldn't be consistent with the rest of our language in the syntax. But I think it's fine.

stepchowfun commented on main.tex:
```diff
@@ -218,17 +232,17 @@
+          \AxiomC{$\xpre{\xunion{\xeffects_3}{\xeffects_4}}{\xsub{\xeffects_2}{\tx_5}{\xeffects_1}}$}
         \RightLabel{(\textsc{C-Arrow})}
-        \TrinaryInfC{$\xusing{\xeffects_1}{\xtapp{\xeffect}{\tx}}{\twithx{\parens{\tarrow{\tx_1}{\tx_2}}}{\xeffects_2}}{\twithx{\parens{\tarrow{\tx_3}{\tx_4}}}{\xeffects_3}}$}
+        \TrinaryInfC{$\xusing{\xeffects_1}{\tx_6}{\twithx{\parens{\tarrow{\tx_1}{\tx_2}}}{\xeffects_2}}{\twithx{\parens{\tarrow{\tx_3}{\tx_4}}}{\xeffects_3}}$}
```
maybe a dumb question but: shouldn't tx5 and tx6 be the same?

ewang12 replied:
Yes

ewang12 commented:
I changed "type with effects" to be "scheme." It's not perfect, but we can always change it again later. I also fixed the issue with `C-Arrow` and reordered a couple blocks of syntax.
